zookeeper-server/src/main/java/org/apache/zookeeper/server/ZooKeeperServerMain.java

```diff
@@ -237,4 +237,26 @@ ServerCnxnFactory getCnxnFactory() {
         return cnxnFactory;
     }
 
+    /**
+     * Shutdowns properly the service, this method is not a public API.
+     */
+    public void close() {
+        ServerCnxnFactory primaryCnxnFactory = this.cnxnFactory;
+        if (primaryCnxnFactory == null) {
+            // in case of pure TLS we can hook into secureCnxnFactory
+            primaryCnxnFactory = secureCnxnFactory;
+        }
+        if (primaryCnxnFactory == null || primaryCnxnFactory.getZooKeeperServer() == null) {
+            return;
+        }
+        ZooKeeperServerShutdownHandler zkShutdownHandler = primaryCnxnFactory.getZooKeeperServer().getZkShutdownHandler();
+        zkShutdownHandler.handle(ZooKeeperServer.State.SHUTDOWN);
```

Comment:
Just curious, and showing my ignorance here: why not:

```java
primaryCnxnFactory.getZooKeeperServer().setState(ZooKeeperServer.State.SHUTDOWN);
```

?

Reply:
ZooKeeperServerShutdownHandler contains a CountDownLatch, that is keeping alive the server itself.
We have to trigger that latch

Reply:
Interesting.  My initial point was that `ZooKeeperServer.setState` invokes `zkShutdownHandler`'s `handle` method—which itself decrements the latch—but I now see that `QuorumZooKeeperServer` and `ReadOnlyZooKeeperServer` override that method, and do not delegate to the superclass.  (Part of the backstory is at [ZOOKEEPER-2247](https://issues.apache.org/jira/browse/ZOOKEEPER-2247), but that's quite a long thread and I haven't deciphered it fully.)